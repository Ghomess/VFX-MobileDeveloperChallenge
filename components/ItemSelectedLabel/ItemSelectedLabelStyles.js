import {StyleSheet} from 'react-native';

export const itemSelectedLabelStyles = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'row',
    width: '100%',
    alignContent: 'center',
    justifyContent: 'space-between',
    padding: 15,
  },

  itemName: {
    color: '#424242',
    fontSize: 20,
    fontWeight: 'bold',
  },
  itemPrice: {
    color: '#1A237E',
    fontSize: 20,
    fontWeight: 'bold',
  },
  itemMonth: {
    color: '#424242',
    fontSize: 20,
    fontWeight: 'bold',
  },
});
