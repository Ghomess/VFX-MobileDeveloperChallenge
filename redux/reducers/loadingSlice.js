import {createSlice} from '@reduxjs/toolkit';

const initialState = {
  loadingChart: false,
  loadingSearchResults: false,
};

const loadingSlice = createSlice({
  name: 'loading',
  initialState,
  reducers: {
    changeLoadingChart: (state, action) => {
      console.log('CHANGE loadingChart: ', action.payload);
      state.loadingChart = action.payload;
    },
    changeLoadingSearchResults: (state, action) => {
      console.log('CHANGE loadingSearchResults: ', action.payload);
      state.loadingSearchResults = action.payload;
    },
  },
});

export const {changeLoadingChart, changeLoadingSearchResults} =
  loadingSlice.actions;

export default loadingSlice.reducer;
