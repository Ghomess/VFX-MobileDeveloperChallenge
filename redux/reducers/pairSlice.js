import {createSlice} from '@reduxjs/toolkit';

const initialState = {
  pair: null,
  pairData: [],
  pairDateSelected: '',
  pairDateType: 'Daily',
};

const pairSlice = createSlice({
  name: 'pair',
  initialState,
  reducers: {
    addPair: (state, action) => {
      console.log('ADD PAIR: ', action.payload);
      state.pair = action.payload;
    },
    addPairData: (state, action) => {
      console.log('ADD PAIR DATA: ', action.payload);
      state.pairData = action.payload;
    },
    addPairDateSelected: (state, action) => {
      console.log('ADD PAIR DATE: ', action.payload);
      state.pairDateSelected = action.payload;
    },
    addPairDateType: (state, action) => {
      console.log('ADD PAIR DATE: ', action.payload);
      state.pairDateType = action.payload;
    },
  },
});

export const {addPair, addPairData, addPairDateSelected, addPairDateType} =
  pairSlice.actions;

export default pairSlice.reducer;
