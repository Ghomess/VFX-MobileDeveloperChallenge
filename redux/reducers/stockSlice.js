import {createSlice} from '@reduxjs/toolkit';

const initialState = {
  ticker: null,
  stockData: {},
  tickerSearchResults: [],
  stockMonthSelected: null,
  stockDateType: 'Monthly',
  stockPrice: null,
};

const stockSlice = createSlice({
  name: 'ticker',
  initialState,
  reducers: {
    addTicker: (state, action) => {
      console.log('ADD ticker: ', action.payload);
      state.ticker = action.payload;
    },
    addStockData: (state, action) => {
      console.log('ADD stock DATA: ', action.payload);
      state.stockData = action.payload;
    },
    addStockMonthSelected: (state, action) => {
      console.log('ADD stock MONTH: ', action.payload);
      state.stockMonthSelected = action.payload;
    },
    addTickerSearchResults: (state, action) => {
      console.log('ADD ticker SEARCH RESULTS: ', action.payload);
      state.tickerSearchResults = action.payload;
    },
    addStockPrice: (state, action) => {
      console.log('ADD stock PRICE: ', action.payload);
      state.stockPrice = action.payload;
    },
  },
});

export const {
  addTicker,
  addStockData,
  addStockMonthSelected,
  addTickerSearchResults,
  addStockPrice,
} = stockSlice.actions;

export default stockSlice.reducer;
