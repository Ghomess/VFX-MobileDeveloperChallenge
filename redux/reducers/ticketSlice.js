import {createSlice} from '@reduxjs/toolkit';

const initialState = {
  ticket: null,
  ticketData: {},
  ticketSearchResults: [],
  ticketMonthSelected: null,
};

const ticketSlice = createSlice({
  name: 'ticket',
  initialState,
  reducers: {
    addTicket: (state, action) => {
      console.log('ADD TICKET: ', action.payload);
      state.ticket = action.payload;
    },
    addTicketData: (state, action) => {
      console.log('ADD TICKET DATA: ', action.payload);
      state.ticketData = action.payload;
    },
    addTicketMonth: (state, action) => {
      console.log('ADD TICKET MONTH: ', action.payload);
      state.ticketMonthSelected = action.payload;
    },
    addTicketSearchResults: (state, action) => {
      console.log('ADD TICKET SEARCH RESULTS: ', action.payload);
      state.ticketSearchResults = action.payload;
    },
  },
});

export const {
  addTicket,
  addTicketData,
  addTicketMonth,
  addTicketSearchResults,
} = ticketSlice.actions;

export default ticketSlice.reducer;
