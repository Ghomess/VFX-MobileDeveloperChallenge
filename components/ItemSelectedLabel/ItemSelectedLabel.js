import React, {useCallback, useState} from 'react';
import {Text, TouchableOpacity, View} from 'react-native';
import {itemSelectedLabelStyles} from './ItemSelectedLabelStyles';
import {useDispatch, useSelector} from 'react-redux';
import {useFocusEffect, useNavigation} from '@react-navigation/native';
import FontAwesome from 'react-native-vector-icons/FontAwesome';
import {colors} from '../../styles';
import {pairDataFetch, stockDataFetch} from '../../utils/apiFunctions';

export function ItemSelectedLabel() {
  //add current price when we have real data using redux
  const navigation = useNavigation();
  const navigationState = navigation.getState();
  const currentScreen = navigationState.routes[navigationState.index].name;
  const dispatch = useDispatch();
  const pair = useSelector(state => state.pair.pair);
  const ticker = useSelector(state => state.ticker.ticker);
  const tickerDateSelected = useSelector(
    state => state.ticker.stockMonthSelected,
  );
  const pairDateSelected = useSelector(state => state.pair.pairDateSelected);
  const pairDateType = useSelector(state => state.pair.pairDateType);
  const stockDateType = useSelector(state => state.ticker.stockDateType);
  const pairPrice = useSelector(state => state.pair.pairPrice);
  const stockPrice = useSelector(state => state.ticker.stockPrice);

  const monthNames = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
  ];

  const [priceSuffix, setPriceSuffix] = useState();
  const checkDataType = () => {
    let screenDateSelected;
    let screenDateType;
    if (currentScreen === 'Stocks') {
      screenDateSelected = tickerDateSelected;
      screenDateType = stockDateType;
    } else if (currentScreen === 'Currency') {
      screenDateSelected = pairDateSelected;
      screenDateType = pairDateType;
    }

    if (screenDateType === 'Monthly') {
      return monthNames[new Date(screenDateSelected).getMonth()];
    } else {
      return screenDateSelected;
    }
  };
  function PriceSuffixFunction() {
    let price;
    if (currentScreen === 'Currency' && pair) {
      price = pairPrice;
      const pairSepareted = pair.split('/');
      if (pairSepareted[0] === 'EUR') {
        console.log('€');
        setPriceSuffix(price + '€');
      } else if (pairSepareted[0] === 'USD') {
        console.log('$');
        setPriceSuffix(price + '$');
      } else if (pairSepareted[0] === 'GBP') {
        console.log('£');
        setPriceSuffix(price + '£');
      } else {
        console.log('$');
        setPriceSuffix(price + '$');
      }
    } else if (currentScreen === 'Stocks' && ticker) {
      price = stockPrice;
      console.log('$');
      setPriceSuffix(price + '$');
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }

  function refresh() {
    if (currentScreen === 'Currency' && pair) {
      pairDataFetch(dispatch, pair, pairDateType);
    } else if (currentScreen === 'Stocks' && ticker) {
      stockDataFetch(dispatch, ticker, stockDateType);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }
  useFocusEffect(
    useCallback(() => {
      PriceSuffixFunction();

      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [currentScreen, pairPrice, stockPrice]),
  );
  //SHOW WEEK NUMBER AND DAY
  return (
    <View style={itemSelectedLabelStyles.container}>
      <Text style={itemSelectedLabelStyles.itemMonth}>{checkDataType()}</Text>
      <View style={itemSelectedLabelStyles.pricerefreshContainer}>
        <Text style={itemSelectedLabelStyles.itemPrice}>{priceSuffix}</Text>
        <TouchableOpacity
          style={itemSelectedLabelStyles.refreshButton}
          onPress={() => refresh()}>
          <FontAwesome name={'refresh'} size={20} color={colors.white} />
        </TouchableOpacity>
      </View>
    </View>
  );
}
