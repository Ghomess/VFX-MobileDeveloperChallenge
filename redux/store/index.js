import {configureStore} from '@reduxjs/toolkit';
import stockSlice from '../reducers/stockSlice';
import pairSlice from '../reducers/pairSlice';
import loadingSlice from '../reducers/loadingSlice';

const store = configureStore({
  reducer: {
    ticker: stockSlice,
    pair: pairSlice,
    loading: loadingSlice,
  },
});

export default store;
