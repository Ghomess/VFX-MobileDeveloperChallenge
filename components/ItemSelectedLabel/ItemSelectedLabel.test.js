import React from 'react';

import {NavigationContainer} from '@react-navigation/native';
import {Provider} from 'react-redux';

import {fireEvent, render, waitFor} from '@testing-library/react-native';
import {ItemSelectedLabel} from './ItemSelectedLabel';
import {changeLoadingChart} from '../../redux/reducers/loadingSlice';
import {configureStore} from '@reduxjs/toolkit';
import stockSlice from '../../redux/reducers/stockSlice';
import pairSlice from '../../redux/reducers/pairSlice';
import store from '../../redux/store';

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => ({
    getState: () => ({
      routes: [{name: 'Stocks'}, {name: 'Currency'}],
      index: 0,
    }),
  }),
}));

describe('ItemSelectedLabel renders correctly', () => {
  test('renders correctly with a currency selected', async () => {
    await waitFor(() => {
      jest
        .spyOn(require('@react-navigation/native'), 'useNavigation')
        .mockImplementation(() => ({
          getState: () => ({
            routes: [{name: 'Stocks'}, {name: 'Currency'}], // Set the routes to only contain the 'Currency' route
            index: 1, // Set the index to 0 to indicate the first route
          }),
          navigate: jest.fn(), // Mock the navigate function to track navigation actions
        }));
      const initialState = {
        pair: {
          pair: 'EUR/USD',
          pairPrice: 1.2345,
          pairDateSelected: '2022-01-01',
          pairDateType: 'Monthly',
        },
      };

      const storeMock = configureStore({
        reducer: {
          pair: pairSlice,
          ticker: stockSlice,
        },
        preloadedState: initialState,
      });

      const tree = render(
        <Provider store={storeMock}>
          <NavigationContainer>
            <ItemSelectedLabel navigation={{}} />
          </NavigationContainer>
        </Provider>,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  test('renders correctly with a stock selected', async () => {
    await waitFor(() => {
      jest
        .spyOn(require('@react-navigation/native'), 'useNavigation')
        .mockImplementation(() => ({
          getState: () => ({
            routes: [{name: 'Stocks'}, {name: 'Currency'}], // Set the routes to only contain the 'Currency' route
            index: 0, // Set the index to 0 to indicate the first route
          }),
          navigate: jest.fn(), // Mock the navigate function to track navigation actions
        }));
      const initialState = {
        ticker: {
          stockDateType: 'Monthly',
          ticker: 'AAPL',
          stockPrice: 100.2345,
          stockMonthSelected: '2022-01-01',
        },
      };

      const storeMock = configureStore({
        reducer: {
          ticker: stockSlice,
          pair: pairSlice,
        },
        preloadedState: initialState,
      });

      const tree = render(
        <Provider store={storeMock}>
          <NavigationContainer>
            <ItemSelectedLabel />
          </NavigationContainer>
        </Provider>,
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});

describe('ItemSelectedLabel dispatches', () => {
  test('check if it dispatches changeLoadingChart action with a stock selected', async () => {
    await waitFor(() => {
      jest
        .spyOn(require('@react-navigation/native'), 'useNavigation')
        .mockImplementation(() => ({
          getState: () => ({
            routes: [{name: 'Stocks'}, {name: 'Currency'}], // Set the routes to only contain the 'Currency' route
            index: 0, // Set the index to 0 to indicate the first route
          }),
          navigate: jest.fn(), // Mock the navigate function to track navigation actions
        }));
      const dispatchMock = jest.spyOn(store, 'dispatch');
      const {getByTestId} = render(
        <Provider store={store}>
          <NavigationContainer>
            <ItemSelectedLabel />
          </NavigationContainer>
        </Provider>,
      );
      const touchable = getByTestId('ItemSelectedLabel.ToucableOpacity');
      fireEvent.press(touchable);
      expect(dispatchMock).toHaveBeenCalledWith(changeLoadingChart(true));
    });
  });

  test('check if it dispatches changeLoadingChart action with a currency selected', async () => {
    await waitFor(() => {
      jest
        .spyOn(require('@react-navigation/native'), 'useNavigation')
        .mockImplementation(() => ({
          getState: () => ({
            routes: [{name: 'Stocks'}, {name: 'Currency'}], // Set the routes to only contain the 'Currency' route
            index: 1, // Set the index to 0 to indicate the first route
          }),
          navigate: jest.fn(), // Mock the navigate function to track navigation actions
        }));
      const dispatchMock = jest.spyOn(store, 'dispatch');
      const {getByTestId} = render(
        <Provider store={store}>
          <NavigationContainer>
            <ItemSelectedLabel />
          </NavigationContainer>
        </Provider>,
      );
      const touchable = getByTestId('ItemSelectedLabel.ToucableOpacity');
      fireEvent.press(touchable);
      expect(dispatchMock).toHaveBeenCalledWith(changeLoadingChart(true));
    });
  });
});
describe('ItemSelectedLabel refresh function with both items selected', () => {
  test('check if it refreshes when currentScreen is Stocks', async () => {
    await waitFor(() => {
      const dispatchMock = jest.spyOn(store, 'dispatch');
      const pairDataFetchMock = jest.fn();
      const stockDataFetchMock = jest.fn();
      const currentScreen = 'Stocks';
      const pair = 'EUR/USD';
      const ticker = 'META';
      const pairDateType = 'Day';
      const stockDateType = 'Monthly';
      // Replace the original pairDataFetch and stockDataFetch functions with mock functions
      const refresh = jest.fn(() => {
        if (currentScreen === 'Currency' && pair) {
          pairDataFetchMock(dispatchMock, pair, pairDateType);
        } else if (currentScreen === 'Stocks' && ticker) {
          stockDataFetchMock(dispatchMock, ticker, stockDateType);
        }
        console.log('refreshing');
        dispatchMock(changeLoadingChart(true));
      });

      // Call the refresh function with the mock arguments
      refresh(dispatchMock, currentScreen, ticker, stockDateType);

      // Assert that pairDataFetch or stockDataFetch was called based on the screen and ticker/pair
      if (currentScreen === 'Currency' && pair) {
        expect(pairDataFetchMock).toHaveBeenCalledWith(
          dispatchMock,
          pair,
          pairDateType,
        );
        expect(stockDataFetchMock).not.toHaveBeenCalled();
      } else if (currentScreen === 'Stocks' && ticker) {
        expect(stockDataFetchMock).toHaveBeenCalledWith(
          dispatchMock,
          ticker,
          stockDateType,
        );
        expect(pairDataFetchMock).not.toHaveBeenCalled();
      }

      // Assert that changeLoadingChart was called with true
      expect(dispatchMock).toHaveBeenCalledWith(changeLoadingChart(true));
    });
  });
  test('check if it refreshes when currentScreen is Currency', async () => {
    const dispatchMock = jest.spyOn(store, 'dispatch');
    const pairDataFetchMock = jest.fn();
    const stockDataFetchMock = jest.fn();
    const currentScreen = 'Currency';
    const pair = 'EUR/USD';
    const ticker = 'META';
    const pairDateType = 'Day';
    const stockDateType = 'Monthly';
    // Replace the original pairDataFetch and stockDataFetch functions with mock functions
    const refresh = jest.fn(() => {
      if (currentScreen === 'Currency' && pair) {
        pairDataFetchMock(dispatchMock, pair, pairDateType);
      } else if (currentScreen === 'Stocks' && ticker) {
        stockDataFetchMock(dispatchMock, ticker, stockDateType);
      }

      dispatchMock(changeLoadingChart(true));
    });

    // Call the refresh function with the mock arguments
    refresh(dispatchMock, currentScreen, pair, pairDateType);

    // Assert that pairDataFetch or stockDataFetch was called based on the screen and ticker/pair
    if (currentScreen === 'Currency' && pair) {
      expect(pairDataFetchMock).toHaveBeenCalledWith(
        dispatchMock,
        pair,
        pairDateType,
      );
      expect(stockDataFetchMock).not.toHaveBeenCalled();
    } else if (currentScreen === 'Stocks' && ticker) {
      expect(stockDataFetchMock).toHaveBeenCalledWith(
        dispatchMock,
        ticker,
        stockDateType,
      );
      expect(pairDataFetchMock).not.toHaveBeenCalled();
    }

    // Assert that changeLoadingChart was called with true
    expect(dispatchMock).toHaveBeenCalledWith(changeLoadingChart(true));
  });
});
