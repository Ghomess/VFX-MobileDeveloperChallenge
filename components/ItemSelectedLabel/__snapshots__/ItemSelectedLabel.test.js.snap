// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemSelectedLabel renders correctly renders correctly with a currency selected 1`] = `
<View
  style={
    {
      "alignContent": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "padding": 15,
      "width": "100%",
    }
  }
>
  <Text
    style={
      {
        "color": "#424242",
        "fontSize": 20,
        "fontWeight": "bold",
      }
    }
  >
    January
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        {
          "color": "#1A237E",
          "fontSize": 20,
          "fontWeight": "bold",
        }
      }
    >
      1.2345€
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#1A237E",
          "borderRadius": 15,
          "marginLeft": 10,
          "opacity": 1,
          "padding": 5,
        }
      }
      testID="ItemSelectedLabel.ToucableOpacity"
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#ffffff",
              "fontSize": 20,
            },
            undefined,
            {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;

exports[`ItemSelectedLabel renders correctly renders correctly with a stock selected 1`] = `
<View
  style={
    {
      "alignContent": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "padding": 15,
      "width": "100%",
    }
  }
>
  <Text
    style={
      {
        "color": "#424242",
        "fontSize": 20,
        "fontWeight": "bold",
      }
    }
  >
    January
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        {
          "color": "#1A237E",
          "fontSize": 20,
          "fontWeight": "bold",
        }
      }
    >
      100.2345$
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#1A237E",
          "borderRadius": 15,
          "marginLeft": 10,
          "opacity": 1,
          "padding": 5,
        }
      }
      testID="ItemSelectedLabel.ToucableOpacity"
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#ffffff",
              "fontSize": 20,
            },
            undefined,
            {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;
